src/vm/callables.rs

```diff
@@ -83,6 +105,38 @@ impl DefinedFunction {
         }
     }
 
+    pub fn check_trait_expectations(&self, 
+                                    contract_defining_trait: &ContractContext,
+                                    trait_identifier: &TraitIdentifier) -> Result<()> {
+
+        let trait_name = trait_identifier.name.to_string();
+        let constraining_trait = contract_defining_trait.lookup_trait_definition(&trait_name)
+            .ok_or(CheckErrors::TraitReferenceUnknown(trait_name.to_string()))?;
+        let expected_sig = constraining_trait.get(&self.name)
+            .ok_or(CheckErrors::TraitMethodUnknown(trait_name.to_string(), self.name.to_string()))?;
+
+        if expected_sig.args.len() != self.arg_types.len() {
+            return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), self.name.to_string()).into())
+        }
+
+        let args = expected_sig.args.iter().zip(self.arg_types.iter());
+        for (expected_arg, actual_arg) in args {
+            match (expected_arg, actual_arg) {
+                (TypeSignature::TraitReferenceType(expected), TypeSignature::TraitReferenceType(ref actual)) => {
+                    if actual != expected {
+                        return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), self.name.to_string()).into())
+                    }
+                }
+                (_, arg_sig) => {
+                    if !expected_arg.admits_type(&arg_sig) {
```

Comment:
Same comment here as in the trait_checker -- expected_arg should be admitted by arg.
Also, can this be refactored to share the same arg checking code as the trait_checker?

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/058549a72fd3e882bcef0b45ccd31c12a55051ee